README.md
TideHunter
===========

HTTP streaming toolbox with flow control.

[![Build Status](https://travis-ci.org/amoa/tidehunter.png?branch=master)](https://travis-ci.org/amoa/tidehunter)

## Highlights

* Stream data queue, state machine, and record counter based on Redis and redis-py.
* Based on the solid cURL and PycURL.
* OAuth supported based on python-oauth2

## Dependencies

`$ pip install oauth2 redis hiredis pycurl`

## Documentation

Coming up very soon!


CHANGES
0.1.1
-----

[Minor]: clean up setup.py to ensure requirements are installed/updated.

0.1.0
-----

Initial release


test/config.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

HUNTER_CONF = {
    'url': 'https://stream.twitter.com/1.1/statuses/sample.json',
    'oauth': {
        'consumer_key': '5wgrdrBSvbfXBuopdGgY9Q',
        'consumer_secret': 'GKtn1buVpvynFlDsFmTl8tRJmnbVsSmULRGUbEMNU',
        'token_key': '974720028-xyfPogLX9ynU05xQYwYNiPtiZengdpdYbyxnpXBI',
        'token_secret': 'rFDXV79FIG3881oxYT3XpYKEXi22bk74DaM7poYR0g'
    },
    'limit': 5
}


test/stream_test.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import unittest
import random
import string

import sys
import os

target_path = os.path.join(os.path.dirname(__file__), '..', 'tidehunter')
sys.path.append(target_path)

# Test Targets
from stream import StateCounter, Queue, Hunter


class HunterTest(unittest.TestCase):

    def _conf(self):
        self.hunter_conf = {
            'url': 'https://httpbin.org/stream/20',
            'limit': 5,
            'delimiter': '\n',
            'timeout': 30
        }

    def setUp(self):
        self.sc_key = 'test_sc'
        self.sc = StateCounter(key=self.sc_key, host='localhost', port=6379,
                               db=0)
        self.q_key = 'test_q'
        self.q = Queue(key=self.q_key, host='localhost', port=6379, db=0)

        self._conf()  # load in the hunter_conf
        self.h = Hunter(self.hunter_conf, self.sc, self.q)

    def test_tide_on(self):
        self.h.tide_on()
        self.assertEqual(self.sc.get_total(), 5)

    def tearDown(self):
        self.sc.conn.delete(self.sc_key)
        self.q.conn.delete(self.q_key)

        try:
            self.h.conn.close()
        except:  # pragma: no cover
            pass


class HunterTestOAuth(HunterTest):

    def _conf(self):
        self.hunter_conf = {
            'url': 'https://stream.twitter.com/1.1/statuses/sample.json',
            'oauth': {
                'consumer_key': '5wgrdrBSvbfXBuopdGgY9Q',
                'consumer_secret': 'GKtn1buVpvynFlDsFmTl8tRJmnbVsSmULRGUbEMNU',
                'token_key': '974720028-xyfPogLX9ynU05xQYwYNiPtiZengdpdYbyxnpXBI',
                'token_secret': 'rFDXV79FIG3881oxYT3XpYKEXi22bk74DaM7poYR0g'
            },
            'limit': 5,
            'timeout': 30
        }


class QueueTest(unittest.TestCase):

    def setUp(self):
        self.key = "test_q"
        self.q = Queue(key=self.key, host='localhost', port=6379, db=0)
        self.q.conn.delete(self.key)

    def test__len__(self):
        # When empty
        self.assertEqual(len(self.q), 0)

        # When not empty
        length = random.randint(1, 32)

        for i in xrange(length):
            self.q.conn.rpush(self.key, i)

        self.assertEqual(len(self.q), length)

    def test_put(self):
        val = ''.join(random.choice(string.ascii_uppercase + string.digits)
                      for x in range(32))
        self.assertTrue(self.q.put(val))
        self.assertEqual(self.q.conn.lpop(self.key), val)

    def test_get(self):
        # When empty
        # self.assertIsNone(self.q.get())
        self.assertTrue(self.q.get() is None)  # 2.1 - 2.6 support

        # When not empty
        val = ''.join(random.choice(string.ascii_uppercase + string.digits)
                      for x in range(32))
        self.q.conn.rpush(self.key, val)
        self.assertEqual(self.q.get(), val)

    def tearDown(self):
        self.q.conn.delete(self.key)


class StateCounterTest(unittest.TestCase):

    def setUp(self):
        self.key = "test_sc"
        self.sc = StateCounter(key=self.key, host='localhost', port=6379,
                               db=0)

    def test_str(self):
        self.assertEqual(str(self.sc), "State: 1\tCount: 0\tTotal: 0")

    def test_get_state(self):
        self.assertEqual(self.sc.get_state(), 1)

    def test_get_count(self):
        self.assertEqual(self.sc.get_count(), 0)

    def test_get_total(self):
        self.assertEqual(self.sc.get_total(), 0)

    def test_get_all(self):
        expected = {'count': '0', 'state': '1', 'total': '0'}
        actual = self.sc.get_all()

        self.assertEqual(actual, expected)

    def test_start(self):
        self.sc.conn.hset(self.key, 'state', 0)
        self.assertEqual(self.sc.get_state(), 0)
        self.sc.conn.hset(self.key, 'count', 100)
        self.assertEqual(self.sc.get_count(), 100)

        self.sc.start()

        self.assertEqual(self.sc.get_state(), 1)
        self.assertEqual(self.sc.get_count(), 0)

    def test_stop(self):
        self.assertEqual(self.sc.get_state(), 1)
        self.sc.conn.hset(self.key, 'count', 100)
        self.assertEqual(self.sc.get_count(), 100)
        self.assertEqual(self.sc.get_total(), 0)

        self.sc.stop()

        self.assertEqual(self.sc.get_state(), 0)
        self.assertEqual(self.sc.get_count(), 0)
        self.assertEqual(self.sc.get_total(), 100)

    def test_incr(self):
        self.assertEqual(self.sc.get_count(), 0)
        self.sc.incr()
        self.assertEqual(self.sc.get_count(), 1)

    def test_started(self):
        self.assertTrue(self.sc.started())
        self.sc.stop()
        self.assertFalse(self.sc.started())

    def test_stopped(self):
        self.assertFalse(self.sc.stopped())
        self.sc.stop()
        self.assertTrue(self.sc.stopped())

    def tearDown(self):
        self.sc.conn.delete(self.key)


if __name__ == '__main__':
    unittest.main()  # pragma: no cover


tidehunter/stream.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import time
import pycurl
import oauth2
import redis


class RedisBase(object):

    def __init__(self, key, host='localhost', port=6379, db=0):
        pool = redis.ConnectionPool(host=host, port=port, db=db)
        self.conn = redis.StrictRedis(connection_pool=pool)
        self.key = key

        self.initialize()

    def initialize(self):
        raise NotImplementedError  # pragma: no cover


class StateCounter(RedisBase):

    '''
    A state counter machine to be used with Hunter based on Redis Hash

    Hash fields:
        state: 1 or 0 (on or off, respectively)
        count: positive int value (current count for limit)
        total: positive int value (total count since init)
    '''

    def __str__(self):
        ret = "State: " + str(self.get_state())
        ret += "\tCount: " + str(self.get_count())
        ret += "\tTotal: " + str(self.get_total())

        return ret

    def initialize(self):
        if not self.conn.exists(self.key):
            self.start()
            self.conn.hset(self.key, 'total', 0)

    def get_state(self):
        val = self.conn.hget(self.key, 'state') or 0

        return int(val)

    def get_count(self):
        val = self.conn.hget(self.key, 'count') or 0

        return int(val)

    def get_total(self):
        val = self.conn.hget(self.key, 'total') or 0

        return int(val)

    def get_all(self):
        return self.conn.hgetall(self.key)

    def start(self):
        state = self.stopped()
        count = self.get_count()

        if state and count:
            self.conn.hincrby(self.key, 'total', count)

        self.conn.hset(self.key, 'state', 1)
        self.conn.hset(self.key, 'count', 0)

    def stop(self):
        self.conn.hset(self.key, 'state', 0)
        self.conn.hincrby(self.key, 'total', self.get_count())
        self.conn.hset(self.key, 'count', 0)

    def incr(self):
        self.conn.hincrby(self.key, 'count', 1)

    def started(self):
        return bool(self.get_state())

    def stopped(self):
        return not self.started()


class Queue(RedisBase):

    '''
    A minimum data queue, based on Redis List
    '''

    def initialize(self):
        pass

    def __len__(self):
        return self.conn.llen(self.key)

    def put(self, var):
        return self.conn.rpush(self.key, var)

    def get(self):
        return self.conn.lpop(self.key)


class Hunter(object):

    def __init__(self, conf, sc, q):
        # stats
        # self.backoff = conf.get('backoff', 1)

        # flow control
        self.buffer = ""
        self.q = q
        self.sc = sc
        self.limit = conf.get('limit', 0)

        # connection
        self.conn = None
        self.url = str(conf['url'])
        self.delimiter = conf.get('delimiter', '\r\n')
        self.timeout = conf.get('timeout', 0)
        self.compressed = conf.get('compressed', True)

        # oauth
        oauth = conf.get('oauth')
        self.consumer = None
        self.token = None

        if oauth:
            self.consumer = oauth2.Consumer(
                key=oauth['consumer_key'], secret=oauth['consumer_secret'])
            self.token = oauth2.Token(
                key=oauth['token_key'], secret=oauth['token_secret'])

        # basic auth
        self.user = ''
        self.pw = ''

        if not self.consumer or not self.token:
            self.user = str(conf.get('user', ''))
            self.pw = str(conf.get('pass', ''))

        # setup connection
        self.setup_conn()

    def setup_conn(self):
        if self.conn:
            try:
                self.conn.close()
            except:  # pragma: no cover
                pass

            self.conn = None

        # reset buffer and record count
        self.buffer = ""

        # init curl
        self.conn = pycurl.Curl()
        self.conn.setopt(pycurl.URL, self.url)

        # Setup auth, oauth or basic
        if self.token and self.consumer:
            auth = get_oauth_header(self.consumer, self.token, self.url)
            self.conn.setopt(pycurl.HTTPHEADER, ['Authorization: %s' % auth])
        elif self.user and self.pw:
            auth = "%s:%s" % (self.user, self.pw)
            self.conn.setopt(pycurl.USERPWD, auth)

        # Other HTTP params
        if self.timeout > 0:
            self.conn.setopt(pycurl.CONNECTTIMEOUT, self.timeout)

        if self.compressed:
            self.conn.setopt(pycurl.ENCODING, 'gzip, deflate')

        # Incoming data handler
        self.conn.setopt(pycurl.WRITEFUNCTION, self._on_data_receive)

    def _on_data_receive(self, data):
        self.buffer += data

        if data.endswith(self.delimiter):
            q_data = self.buffer.strip()
            self.buffer = ""

            for j in iter_chunk(q_data, self.delimiter):
                self.q.put(j)

                if self.sc.started():
                    self.sc.incr()

                    if self.limit and self.sc.get_count() >= self.limit:
                        self.sc.stop()

        # TODO: clean up the most recent buffer?
        # currently just disregard it
        if self.sc.stopped():
            return -1  # the way to stop the stream, would raise pycurl.error

    def tide_on(self):
        # while self.sc.started():
        try:
            self.sc.start()
            self.q.put("{}")  # just to reactivate queue
            self.setup_conn()
            self.conn.perform()
        except pycurl.error as e:
            if e[0] == 23:
                pass
            else:
                raise
        except:
            raise
        finally:
            # resets
            self.sc.stop()
            self.buffer = ""

        return self.conn.getinfo(pycurl.HTTP_CODE)


def get_oauth_header(consumer, token, url, method='GET'):
    """ Create and return OAuth header.
    """
    params = {'oauth_version': '1.0',
              'oauth_nonce': oauth2.generate_nonce(),
              'oauth_timestamp': int(time.time())}

    r = oauth2.Request(method=method, parameters=params, url=url)
    r.sign_request(oauth2.SignatureMethod_HMAC_SHA1(), consumer, token)

    return r.to_header()['Authorization'].encode('utf-8')


def iter_chunk(data, delimiter):
    jstr_list = data.split(delimiter)

    for jstr in jstr_list:
        jstr = jstr.strip()

        if jstr:
            yield jstr


demo/limit_per_iter.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import os
import json

# Take the following two lines out if you installed tidehunter through pip
target_path = os.path.join(os.path.dirname(__file__), '..')
sys.path.append(target_path)

from tidehunter.stream import StateCounter, Queue, Hunter


if __name__ == '__main__':
    # Init StateCounter
    key_sc = 'demo_sc'
    sc = StateCounter(key=key_sc, host='localhost', port=6379, db=0)

    # Init Queue
    key_q = 'demo_q'
    q = Queue(key=key_q, host='localhost', port=6379, db=0)

    # Init Hunter with conf
    conf = {
        'url': 'https://httpbin.org/stream/20',
        'limit': 5,  # desired limit, might go over
        'delimiter': '\n'
    }
    h = Hunter(conf=conf, sc=sc, q=q)

    for i in xrange(5):
        try:
            http_code = h.tide_on()

            assert http_code == 200
            print "Iteration " + str(i + 1) + ":\t",

            count = 0

            while len(q):
                data = json.loads(q.get())

                if data:
                    count += 1
                    print count,

            print
        except:
            raise
        finally:
            # Some cleanup
            sc.conn.delete(key_sc)
            q.conn.delete(key_q)


demo/five_tweets.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import os
import json

# Take the following two lines out if you installed tidehunter through pip
target_path = os.path.join(os.path.dirname(__file__), '..')
sys.path.append(target_path)

from tidehunter.stream import StateCounter, Queue, Hunter


if __name__ == '__main__':
    # Init StateCounter
    key_sc = 'demo_sc'
    sc = StateCounter(key=key_sc, host='localhost', port=6379, db=0)

    # Init Queue
    key_q = 'demo_q'
    q = Queue(key=key_q, host='localhost', port=6379, db=0)

    # Init Hunter with conf
    conf = {
        'url': 'https://stream.twitter.com/1.1/statuses/sample.json',
        'oauth': {
            'consumer_key': '5wgrdrBSvbfXBuopdGgY9Q',
            'consumer_secret': 'GKtn1buVpvynFlDsFmTl8tRJmnbVsSmULRGUbEMNU',
            'token_key': '974720028-xyfPogLX9ynU05xQYwYNiPtiZengdpdYbyxnpXBI',
            'token_secret': 'rFDXV79FIG3881oxYT3XpYKEXi22bk74DaM7poYR0g'
        },
        'limit': 5
    }
    h = Hunter(conf=conf, sc=sc, q=q)

    try:
        http_code = h.tide_on()

        assert http_code == 200

        while len(q):
            data = json.loads(q.get())

            if data:
                print(data)
    except:
        raise
    finally:
        # Some cleanup
        sc.conn.delete(key_sc)
        q.conn.delete(key_q)
